pkg/pgclient/client.go

```diff
@@ -33,84 +34,108 @@ type LockFunc = func(ctx context.Context, conn *pgx.Conn) error
+		writerPoolSize, err = cfg.GetPoolSize("writer", 0.5, cfg.WriterPoolSize)
+		if err != nil {
+			return nil, fmt.Errorf("get writer pool size: %w", err)
+		}
+		numCopiers, err = cfg.GetNumCopiers()
+		if err != nil {
+			return nil, fmt.Errorf("get num copiers: %w", err)
+		}
+
+		if numCopiers >= writerPoolSize {
+			log.Warn("msg", "number of copiers greater than the writer-pool. Decreasing copiers to leave some connections for miscellaneous write tasks")
+			numCopiers /= 2
+		}
+
+		writerPgConfig, err := cfg.getPgConfig(writerPoolSize)
+		if err != nil {
+			return nil, fmt.Errorf("get writer pg-config: %w", err)
+		}
+		writerPgConfig.AfterConnect = schemaLocker
+		writerPool, err = pgxpool.ConnectConfig(context.Background(), writerPgConfig)
+		if err != nil {
+			return nil, fmt.Errorf("err creating writer connection pool: %w", err)
+		}
+	}
+
+	readerPoolSize, err := cfg.GetPoolSize("reader", defaultReaderPool, cfg.ReaderPoolSize)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("get reader pool size: %w", err)
+	}
+	readerPgConfig, err := cfg.getPgConfig(readerPoolSize)
+	if err != nil {
+		return nil, fmt.Errorf("get reader pg-config: %w", err)
 	}
 
-	pgConfig.AfterConnect = schemaLocker
-	connectionPool, err := pgxpool.ConnectConfig(context.Background(), pgConfig)
+	statementCacheLog := "disabled"
+	if cfg.EnableStatementsCache {
+		statementCacheLog = "512" // Default pgx.
+	}
+	log.Info("msg", getRedactedConnStr(cfg.GetConnectionStr()))
+	log.Info("msg", "runtime",
+		"writer-pool.size", writerPoolSize,
+		"reader-pool.size", readerPoolSize,
+		"min-pool.connections", MinPoolSize,
+		"num-copiers", numCopiers,
+		"statement-cache", statementCacheLog)
+
+	readerPgConfig.AfterConnect = schemaLocker
+	readerPool, err := pgxpool.ConnectConfig(context.Background(), readerPgConfig)
 	if err != nil {
-		log.Error("msg", "err creating connection pool for new client", "err", err.Error())
-		return nil, err
+		return nil, fmt.Errorf("err creating reader connection pool: %w", err)
 	}
 
-	client, err := NewClientWithPool(cfg, numCopiers, connectionPool, mt, readOnly)
+	client, err := NewClientWithPool(r, cfg, numCopiers, writerPool, readerPool, mt, readOnly)
 	if err != nil {
 		return client, err
 	}
-	client.closePool = true
 	return client, err
 }
 
-func getPgConfig(cfg *Config) (*pgxpool.Config, int, error) {
-	minConnections, maxConnections, numCopiers, err := cfg.GetNumConnections()
-	if err != nil {
-		log.Error("msg", "configuring number of connections", "err", err.Error())
-		return nil, numCopiers, err
-	}
-	connectionStr := cfg.GetConnectionStr()
-	pgConfig, err := pgxpool.ParseConfig(connectionStr)
+func (cfg *Config) getPgConfig(poolSize int) (*pgxpool.Config, error) {
+	min := MinPoolSize
+	max := poolSize
+	connStr := cfg.GetConnectionStr()
+
+	pgConfig, err := pgxpool.ParseConfig(connStr)
 	if err != nil {
 		log.Error("msg", "configuring connection", "err", err.Error())
```

Comment:
Should we log the error here? Wouldn't it be also logged by the `runner.Run` function?